app/src/services/bucket.js

```diff
@@ -0,0 +1,170 @@
+   * @function create
+   * Create a bucket record and give the uploader (if authed) permissions
+   * @param {string} data.bucketId The bucket uuid
+   * @param {string} data.bucketName The user-defined bucket name identifier
+   * @param {string} data.accessKeyId The S3 bucket access key id
+   * @param {string} data.bucket The S3 bucket identifier
+   * @param {string} data.endpoint The S3 bucket endpoint
+   * @param {string} data.key The relative S3 key/subpath managed by this bucket
+   * @param {string} data.secretAccessKey The S3 bucket secret access key
+   * @param {boolean} [data.active] The optional active flag - defaults to true if undefined
+   * @param {object} [etrx=undefined] An optional Objection Transaction object
+   * @returns {Promise<object>} The result of running the insert operation
+   * @throws The error encountered upon db transaction failure
+   */
+  create: async (data, etrx = undefined) => {
+    let trx;
+    try {
+      trx = etrx ? etrx : await Bucket.startTransaction();
+
+      // Add bucket record to DB
+      const obj = {
+        bucketId: data.bucketId,
+        bucketName: data.bucketName,
+        accessKeyId: data.accessKeyId,
+        bucket: data.bucket,
+        endpoint: data.endpoint,
+        key: data.key,
+        secretAccessKey: data.secretAccessKey,
+        active: data.active,
+        createdBy: data.userId
+      };
+
+      // TODO: Add conflict/merge logic to avoid bucket duplicates?
+
+      const response = await Bucket.query(trx).insert(obj).returning('*');
+
+      // TODO: Add permissions when all fields match?
+      // Add all permission codes for the uploader
+      // if (data.userId) {
+      //   const perms = Bucket.values(Permissions).map((p) => ({
+      //     userId: data.userId,
+      //     permCode: p
+      //   }));
+      //   await permissionService.addPermissions(data.id, perms, data.userId, trx);
+      // }
+
+      if (!etrx) await trx.commit();
+      return Promise.resolve(response);
+    } catch (err) {
+      if (!etrx && trx) await trx.rollback();
+      throw err;
+    }
+  },
+
+  /**
+   * @function delete
+   * Delete a bucket record. Note this will also delete all objects and permissions
+   * related to this specific bucket from the database. Use with caution.
+   * @param {string} bucketId The bucket uuid to delete
+   * @param {object} [etrx=undefined] An optional Objection Transaction object
+   * @returns {Promise<object>} The result of running the delete operation
+   * @throws The error encountered upon db transaction failure
+   */
+  delete: async (bucketId, etrx = undefined) => {
+    let trx;
+    try {
+      trx = etrx ? etrx : await Bucket.startTransaction();
+
+      const response = await Bucket.query(trx)
+        .deleteById(bucketId)
+        .throwIfNotFound()
+        // Returns array of deleted rows instead of count
+        // https://vincit.github.io/objection.js/recipes/returning-tricks.html
+        .returning('*');
```

Comment:
I want to avoid unnecessary business logic for "soft deletes" at this time as we haven't formally decided what "active" means for an object either. For now this is highly unlikely to be invoked as an endpoint for now, but when there is a delete operation, we do need to drop all children records from the database as there is no more reason for COMS to track dangling objects. At some point in the near future we will need to implement a bucket scan operation that allows you to rehydrate objects into a database, but that's a future discussion on expected behaviors.

Reply:
ok. i guess we can communicate the effects and risks of deleting a bucket